File: modules/mpv-player/ios/NativePlayer/Views/PlayerTopBar.swift
#if os(iOS)
import SwiftUI

/// Top bar: close, title block, then PiP / episode list / speed / audio /
/// subtitle / more menus. Menu contents are display models pushed down by JS
/// (updateTrackMenus); taps emit intent events — selection logic stays in JS.
struct PlayerTopBar: View {
	@ObservedObject var viewModel: PlayerViewModel
	/// Narrow widths (portrait phones): the fixed button row would outgrow
	/// the screen once late data pushes add buttons (episode list), inflating
	/// the whole overlay — so zoom + speed fold into the More menu and the
	/// buttons tighten.
	var compact = false

	private var buttonSize: CGFloat { compact ? 40 : 44 }

	var body: some View {
		HStack(spacing: 4) {
			barButton(systemName: "xmark") {
				viewModel.close()
			}
			.accessibilityLabel(viewModel.str("close", "Close"))

			VStack(alignment: .leading, spacing: 2) {
				if let title = viewModel.metadata?.title {
					Text(title)
						.font(.subheadline.weight(.semibold))
						.foregroundStyle(.white)
						.lineLimit(1)
				}
				if let subtitle = viewModel.metadata?.subtitle {
					Text(subtitle)
						.font(.caption)
						.foregroundStyle(.white.opacity(0.7))
						.lineLimit(1)
				}
			}
			.padding(.leading, 8)

			Spacer(minLength: 12)

			if viewModel.engine?.isPictureInPictureSupported() == true {
				barButton(systemName: viewModel.isPipActive ? "pip.exit" : "pip.enter") {
					viewModel.togglePictureInPicture()
				}
			}

			if !compact {
				barButton(
					systemName: viewModel.isZoomedToFill
						? "arrow.down.right.and.arrow.up.left"
						: "arrow.up.left.and.arrow.down.right"
				) {
					viewModel.toggleZoomToFill()
				}
			}

			if !viewModel.episodeList.isEmpty {
				barButton(systemName: "list.bullet") {
					viewModel.showEpisodeList = true
					viewModel.scheduleAutoHide()
				}
				.accessibilityLabel(viewModel.str("episodes", "Episodes"))
			}

			if !compact {
				speedMenu
			}

			if !viewModel.audioMenu.isEmpty {
				audioTracksMenu
			}

			if !viewModel.subtitleMenu.isEmpty {
				subtitlesMenu
			}

			moreMenu
		}
	}

	// MARK: - Menus

	private var speedMenu: some View {
		Menu {
			speedMenuEntries
		} label: {
			barIcon(systemName: "speedometer")
		}
		.menuOrder(.fixed)
		.simultaneousGesture(TapGesture().onEnded { viewModel.menuInteractionStarted() })
		.accessibilityLabel(viewModel.str("speed", "Speed"))
	}

	private var speedMenuEntries: some View {
		ForEach([0.25, 0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0], id: \.self) { option in
			Button {
				viewModel.setSpeed(option)
			} label: {
				if abs(viewModel.speed - option) < 0.001 {
					Label(speedLabel(option), systemImage: "checkmark")
				} else {
					Text(speedLabel(option))
				}
			}
		}
	}

	/// Audio track picker plus a nested sync-offset submenu (mpv audio-delay).
	private var audioTracksMenu: some View {
		Menu {
			trackMenuItems(viewModel.audioMenu) { viewModel.selectAudio($0) }
			Divider()
			Menu {
				syncOffsetEntries(current: viewModel.audioDelay) { viewModel.setAudioDelay($0) }
			} label: {
				Label(viewModel.str("audioSync", "Audio sync"), systemImage: "clock.arrow.2.circlepath")
			}
		} label: {
			barIcon(systemName: "waveform")
		}
		.menuOrder(.fixed)
		.simultaneousGesture(TapGesture().onEnded { viewModel.menuInteractionStarted() })
		.accessibilityLabel(viewModel.str("audio", "Audio"))
	}

	private func syncOffsetEntries(
		current: Double,
		onSelect: @escaping (Double) -> Void
	) -> some View {
		ForEach(PlayerViewModel.syncOffsetPresets, id: \.self) { offset in
			Button {
				onSelect(offset)
			} label: {
				if abs(current - offset) < 0.001 {
					Label(offsetLabel(offset), systemImage: "checkmark")
				} else {
					Text(offsetLabel(offset))
				}
			}
		}
	}

	private func offsetLabel(_ offset: Double) -> String {
		offset == 0 ? "0 s" : String(format: "%+g s", offset)
	}

	private func trackMenuItems(
		_ items: [TrackMenuItemRecord],
		onSelect: @escaping (TrackMenuItemRecord) -> Void
	) -> some View {
		ForEach(Array(items.enumerated()), id: \.offset) { _, item in
			Button {
				onSelect(item)
			} label: {
				if item.selected {
					Label(item.label, systemImage: "checkmark")
				} else {
					Text(item.label)
				}
			}
		}
	}

	/// Subtitle track picker plus the shared subtitle-size control.
	private var subtitlesMenu: some View {
		Menu {
			trackMenuItems(viewModel.subtitleMenu) { viewModel.selectSubtitle($0) }
			Divider()
			Button {
				viewModel.showSubtitleScaleControl = true
			} label: {
				Label(
					"\(viewModel.str("subtitleSize", "Subtitle size")): \(String(format: "%.1f×", viewModel.subtitleScale))",
					systemImage: "textformat.size"
				)
			}
			.disabled(viewModel.subtitleScaleLocked)
			Menu {
				syncOffsetEntries(current: viewModel.subtitleDelay) { viewModel.setSubtitleDelay($0) }
			} label: {
				Label(viewModel.str("subtitleSync", "Subtitle sync"), systemImage: "clock.arrow.2.circlepath")
			}
		} label: {
			barIcon(systemName: "captions.bubble")
		}
		.menuOrder(.fixed)
		.simultaneousGesture(TapGesture().onEnded { viewModel.menuInteractionStarted() })
		.accessibilityLabel(viewModel.str("subtitles", "Subtitles"))
	}

	private var moreMenu: some View {
		Menu {
			if compact {
				Button {
					viewModel.toggleZoomToFill()
				} label: {
					if viewModel.isZoomedToFill {
						Label(viewModel.str("zoomToFill", "Zoom to fill"), systemImage: "checkmark")
					} else {
						Label(viewModel.str("zoomToFill", "Zoom to fill"), systemImage: "arrow.up.left.and.arrow.down.right")
					}
				}
				Menu {
					speedMenuEntries
				} label: {
					Label(viewModel.str("speed", "Speed"), systemImage: "speedometer")
				}
			}
			if !viewModel.qualityMenu.isEmpty {
				Menu {
					trackMenuItems(viewModel.qualityMenu) { viewModel.selectQuality($0) }
				} label: {
					Label(viewModel.str("quality", "Quality"), systemImage: "slider.horizontal.3")
				}
			}
			if !viewModel.chapters.isEmpty {
				Menu {
					ForEach(Array(viewModel.chapters.enumerated()), id: \.offset) { index, chapter in
						Button {
							viewModel.seek(to: chapter.startSec)
						} label: {
							Text("\(chapter.name ?? "\(index + 1)") — \(formatTime(chapter.startSec))")
						}
					}
				} label: {
					Label(viewModel.str("chapters", "Chapters"), systemImage: "bookmark")
				}
			}
			Menu {
				ForEach(PlayerViewModel.volumeBoostPresets, id: \.self) { percent in
					Button {
						viewModel.setVolumeBoost(percent)
					} label: {
						if viewModel.volumeBoostPercent == percent {
							Label(boostLabel(percent), systemImage: "checkmark")
						} else {
							Text(boostLabel(percent))
						}
					}
				}
			} label: {
				Label(viewModel.str("volumeBoost", "Volume boost"), systemImage: "speaker.wave.3")
			}
			sleepTimerMenu
			Button {
				viewModel.requestRotate()
			} label: {
				Label(viewModel.str("rotate", "Rotate"), systemImage: "rotate.right")
			}
			Button {
				viewModel.lockControls()
			} label: {
				Label(viewModel.str("lockControls", "Lock controls"), systemImage: "lock")
			}
			Button {
				viewModel.showTechnicalInfo.toggle()
				viewModel.scheduleAutoHide()
			} label: {
				if viewModel.showTechnicalInfo {
					Label(viewModel.str("technicalInfo", "Technical info"), systemImage: "checkmark")
				} else {
					Label(viewModel.str("technicalInfo", "Technical info"), systemImage: "info.circle")
				}
			}
		} label: {
			barIcon(systemName: "ellipsis.circle")
		}
		.menuOrder(.fixed)
		.simultaneousGesture(TapGesture().onEnded { viewModel.menuInteractionStarted() })
	}

	/// When the timer fires the player dismisses fully, so the idle timer
	/// re-enables and the phone auto-locks on its own.
	private var sleepTimerMenu: some View {
		Menu {
			if viewModel.sleepTimerMinutes != nil {
				Button {
					viewModel.cancelSleepTimer()
				} label: {
					Label(viewModel.str("sleepTimerOff", "Turn off"), systemImage: "xmark")
				}
				Divider()
			}
			ForEach(PlayerViewModel.sleepTimerPresetMinutes, id: \.self) { minutes in
				Button {
					viewModel.setSleepTimer(minutes: minutes)
				} label: {
					if viewModel.sleepTimerMinutes == minutes {
						Label(sleepTimerLabel(minutes), systemImage: "checkmark")
					} else {
						Text(sleepTimerLabel(minutes))
					}
				}
			}
		} label: {
			Label(
				viewModel.str("sleepTimer", "Sleep timer"),
				systemImage: viewModel.sleepTimerMinutes == nil ? "moon.zzz" : "moon.zzz.fill"
			)
		}
	}

	private func sleepTimerLabel(_ minutes: Int) -> String {
		minutes < 60 ? "\(minutes) min" : String(format: "%g h", Double(minutes) / 60)
	}

	private func speedLabel(_ speed: Double) -> String {
		speed == 1.0 ? "1×" : String(format: "%g×", speed)
	}

	private func boostLabel(_ percent: Int) -> String {
		percent == 100 ? viewModel.str("off", "Off") : "\(percent)%"
	}

	// MARK: - Buttons

	private func barButton(systemName: String, action: @escaping () -> Void) -> some View {
		Button(action: action) {
			barIcon(systemName: systemName)
		}
	}

	private func barIcon(systemName: String) -> some View {
		Image(systemName: systemName)
			.font(.system(size: 17, weight: .semibold))
			.foregroundStyle(.white)
			.frame(width: buttonSize, height: buttonSize)
			.contentShape(Rectangle())
	}
}
#endif

File: modules/mpv-player/android/src/main/java/expo/modules/mpvplayer/nativeplayer/ui/PlayerTopBar.kt
package expo.modules.mpvplayer.nativeplayer.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AspectRatio
import androidx.compose.material.icons.filled.Bookmark
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.ClosedCaption
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.FormatListBulleted
import androidx.compose.material.icons.filled.GraphicEq
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.Nightlight
import androidx.compose.material.icons.filled.PictureInPicture
import androidx.compose.material.icons.filled.RotateRight
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Speed
import androidx.compose.material.icons.filled.Sync
import androidx.compose.material.icons.filled.TextFormat
import androidx.compose.material.icons.filled.VolumeUp
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import expo.modules.mpvplayer.nativeplayer.PlayerConstants
import expo.modules.mpvplayer.nativeplayer.PlayerViewModel
import expo.modules.mpvplayer.nativeplayer.TrackMenuItemRecord

@Composable
fun PlayerTopBar(
    viewModel: PlayerViewModel,
    compact: Boolean = false,
    onPipClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    var showSpeedMenu by remember { mutableStateOf(false) }
    var showAudioMenu by remember { mutableStateOf(false) }
    var showSubtitleMenu by remember { mutableStateOf(false) }
    var showMoreMenu by remember { mutableStateOf(false) }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 12.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Close button
        IconButton(
            onClick = { viewModel.close() },
            modifier = Modifier.size(40.dp)
        ) {
            Icon(
                imageVector = Icons.Default.Close,
                contentDescription = viewModel.strings.get("close", "Close"),
                tint = Color.White
            )
        }

        // Title and Subtitle
        Column(
            modifier = Modifier
                .weight(1f)
                .padding(horizontal = 8.dp)
        ) {
            viewModel.metadata?.title?.let { title ->
                Text(
                    text = title,
                    color = Color.White,
                    fontWeight = FontWeight.SemiBold,
                    fontSize = 15.sp,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
            viewModel.metadata?.subtitle?.let { subtitle ->
                Text(
                    text = subtitle,
                    color = Color.White.copy(alpha = 0.7f),
                    fontSize = 12.sp,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }

        // PiP Button
        if (onPipClick != null) {
            IconButton(
                onClick = { onPipClick() },
                modifier = Modifier.size(40.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.PictureInPicture,
                    contentDescription = "Picture in Picture",
                    tint = Color.White
                )
            }
        }

        // Zoom to fill button (if not compact)
        if (!compact) {
            IconButton(
                onClick = { viewModel.toggleZoomToFill() },
                modifier = Modifier.size(40.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.AspectRatio,
                    contentDescription = viewModel.strings.get("zoomToFill", "Zoom to Fill"),
                    tint = if (viewModel.isZoomedToFill) PlayerAccentColor else Color.White
                )
            }
        }

        // Episode list button
        if (viewModel.episodeList.isNotEmpty()) {
            IconButton(
                onClick = {
                    viewModel.showEpisodeList = true
                    viewModel.scheduleAutoHide()
                },
                modifier = Modifier.size(40.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.FormatListBulleted,
                    contentDescription = viewModel.strings.get("episodes", "Episodes"),
                    tint = Color.White
                )
            }
        }

        // Speed Menu Button (if not compact)
        if (!compact) {
            Box {
                IconButton(
                    onClick = {
                        showSpeedMenu = true
                        viewModel.menuInteractionStarted()
                    },
                    modifier = Modifier.size(40.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.Speed,
                        contentDescription = viewModel.strings.get("speed", "Speed"),
                        tint = Color.White
                    )
                }

                DropdownMenu(
                    expanded = showSpeedMenu,
                    onDismissRequest = {
                        showSpeedMenu = false
                        viewModel.scheduleAutoHide()
                    }
                ) {
                    PlayerConstants.SPEED_PRESETS.forEach { opt ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(if (opt == 1.0) "1×" else "${opt}×")
                                    if (kotlin.math.abs(viewModel.speed - opt) < 0.01) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.setSpeed(opt)
                                showSpeedMenu = false
                            }
                        )
                    }
                }
            }
        }

        // Audio Menu
        if (viewModel.audioMenu.isNotEmpty()) {
            Box {
                IconButton(
                    onClick = {
                        showAudioMenu = true
                        viewModel.menuInteractionStarted()
                    },
                    modifier = Modifier.size(40.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.GraphicEq,
                        contentDescription = viewModel.strings.get("audio", "Audio"),
                        tint = Color.White
                    )
                }

                DropdownMenu(
                    expanded = showAudioMenu,
                    onDismissRequest = {
                        showAudioMenu = false
                        viewModel.scheduleAutoHide()
                    }
                ) {
                    Text(
                        text = viewModel.strings.get("audio", "Audio Tracks"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )

                    viewModel.audioMenu.forEach { item ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(item.label)
                                    if (item.selected) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.selectAudio(item)
                                showAudioMenu = false
                            }
                        )
                    }

                    HorizontalDivider()

                    // Sync Delays
                    Text(
                        text = viewModel.strings.get("audioSync", "Audio Sync"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )
                    PlayerConstants.SYNC_OFFSET_PRESETS.forEach { offset ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(if (offset == 0.0) "0 s" else "${offset} s")
                                    if (kotlin.math.abs(viewModel.audioDelay - offset) < 0.01) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.setAudioDelay(offset)
                                showAudioMenu = false
                            }
                        )
                    }

                    HorizontalDivider()

                    // Volume Boost
                    Text(
                        text = viewModel.strings.get("volumeBoost", "Volume Boost"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )
                    PlayerConstants.VOLUME_BOOST_PRESETS.forEach { percent ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(if (percent == 100) viewModel.strings.get("off", "Off") else "$percent%")
                                    if (viewModel.volumeBoostPercent == percent) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.setVolumeBoost(percent)
                                showAudioMenu = false
                            }
                        )
                    }

                    HorizontalDivider()

                    // Dialogue Boost
                    DropdownMenuItem(
                        text = {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Text(viewModel.strings.get("dialogueBoost", "Dialogue Boost"))
                                if (viewModel.dialogueBoostEnabled) {
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                }
                            }
                        },
                        onClick = {
                            viewModel.toggleDialogueBoost()
                            showAudioMenu = false
                        }
                    )

                    // Mono Downmix
                    DropdownMenuItem(
                        text = {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Text(viewModel.strings.get("monoAudio", "Mono Audio"))
                                if (viewModel.monoAudioEnabled) {
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                }
                            }
                        },
                        onClick = {
                            viewModel.toggleMonoAudio()
                            showAudioMenu = false
                        }
                    )
                }
            }
        }

        // Subtitles Menu
        if (viewModel.subtitleMenu.isNotEmpty()) {
            Box {
                IconButton(
                    onClick = {
                        showSubtitleMenu = true
                        viewModel.menuInteractionStarted()
                    },
                    modifier = Modifier.size(40.dp)
                ) {
                    Icon(
                        imageVector = Icons.Default.ClosedCaption,
                        contentDescription = viewModel.strings.get("subtitles", "Subtitles"),
                        tint = Color.White
                    )
                }

                DropdownMenu(
                    expanded = showSubtitleMenu,
                    onDismissRequest = {
                        showSubtitleMenu = false
                        viewModel.scheduleAutoHide()
                    }
                ) {
                    Text(
                        text = viewModel.strings.get("subtitles", "Subtitles"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )

                    viewModel.subtitleMenu.forEach { item ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(item.label)
                                    if (item.selected) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.selectSubtitle(item)
                                showSubtitleMenu = false
                            }
                        )
                    }

                    HorizontalDivider()

                    if (!viewModel.subtitleScaleLocked) {
                        Text(
                            text = viewModel.strings.get("subtitleSize", "Subtitle Size"),
                            fontWeight = FontWeight.Bold,
                            fontSize = 13.sp,
                            color = Color.Gray,
                            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                        )
                        DropdownMenuItem(
                            text = { Text("${viewModel.subtitleScale}×") },
                            onClick = {
                                showSubtitleMenu = false
                                viewModel.subtitleScaleOverlayActivity++
                                viewModel.showSubtitleScaleOverlay = true
                            }
                        )
                        HorizontalDivider()
                    }

                    // Subtitle Sync
                    Text(
                        text = viewModel.strings.get("subtitleSync", "Subtitle Sync"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )
                    PlayerConstants.SYNC_OFFSET_PRESETS.forEach { offset ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(if (offset == 0.0) "0 s" else "${offset} s")
                                    if (kotlin.math.abs(viewModel.subtitleDelay - offset) < 0.01) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.setSubtitleDelay(offset)
                                showSubtitleMenu = false
                            }
                        )
                    }

                    if (viewModel.uiOptions.subtitleSearchEnabled) {
                        HorizontalDivider()
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Icon(Icons.Default.Search, contentDescription = null, modifier = Modifier.size(16.dp))
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Text(viewModel.strings.get("searchSubtitles", "Search Subtitles"))
                                }
                            },
                            onClick = {
                                showSubtitleMenu = false
                                viewModel.openSubtitleSearch()
                            }
                        )
                    }
                }
            }
        }

        // More Menu
        Box {
            IconButton(
                onClick = {
                    showMoreMenu = true
                    viewModel.menuInteractionStarted()
                },
                modifier = Modifier.size(40.dp)
            ) {
                Icon(
                    imageVector = Icons.Default.MoreVert,
                    contentDescription = "More",
                    tint = Color.White
                )
            }

            DropdownMenu(
                expanded = showMoreMenu,
                onDismissRequest = {
                    showMoreMenu = false
                    viewModel.scheduleAutoHide()
                }
            ) {
                if (compact) {
                    DropdownMenuItem(
                        text = { Text(viewModel.strings.get("zoomToFill", "Zoom to Fill")) },
                        onClick = {
                            viewModel.toggleZoomToFill()
                            showMoreMenu = false
                        }
                    )
                    HorizontalDivider()
                }

                // Quality options
                if (viewModel.qualityMenu.isNotEmpty()) {
                    Text(
                        text = viewModel.strings.get("quality", "Quality"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )
                    viewModel.qualityMenu.forEach { q ->
                        DropdownMenuItem(
                            text = {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(q.label)
                                    if (q.selected) {
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                    }
                                }
                            },
                            onClick = {
                                viewModel.selectQuality(q)
                                showMoreMenu = false
                            }
                        )
                    }
                    HorizontalDivider()
                }

                // Chapters
                if (viewModel.chapters.isNotEmpty()) {
                    Text(
                        text = viewModel.strings.get("chapters", "Chapters"),
                        fontWeight = FontWeight.Bold,
                        fontSize = 13.sp,
                        color = Color.Gray,
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                    )
                    viewModel.chapters.forEachIndexed { i, ch ->
                        DropdownMenuItem(
                            text = { Text("${ch.name ?: "${i + 1}"} — ${formatTimeSec(ch.startSec)}") },
                            onClick = {
                                viewModel.seekTo(ch.startSec)
                                showMoreMenu = false
                            }
                        )
                    }
                    HorizontalDivider()
                }

                // Sleep Timer
                Text(
                    text = viewModel.strings.get("sleepTimer", "Sleep Timer"),
                    fontWeight = FontWeight.Bold,
                    fontSize = 13.sp,
                    color = Color.Gray,
                    modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp)
                )
                if (viewModel.sleepTimerMinutes != null) {
                    DropdownMenuItem(
                        text = { Text(viewModel.strings.get("sleepTimerOff", "Turn Off")) },
                        onClick = {
                            viewModel.cancelSleepTimer()
                            showMoreMenu = false
                        }
                    )
                }
                PlayerConstants.SLEEP_TIMER_PRESET_MINUTES.forEach { mins ->
                    DropdownMenuItem(
                        text = {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Text(if (mins < 60) "$mins min" else "${mins / 60} h")
                                if (viewModel.sleepTimerMinutes == mins) {
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Icon(Icons.Default.Check, contentDescription = null, modifier = Modifier.size(16.dp))
                                }
                            }
                        },
                        onClick = {
                            viewModel.setSleepTimer(mins)
                            showMoreMenu = false
                        }
                    )
                }

                HorizontalDivider()

                // Rotate
                DropdownMenuItem(
                    text = {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.RotateRight, contentDescription = null, modifier = Modifier.size(16.dp))
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(viewModel.strings.get("rotate", "Rotate"))
                        }
                    },
                    onClick = {
                        viewModel.requestRotate()
                        showMoreMenu = false
                    }
                )

                // Lock Controls
                DropdownMenuItem(
                    text = {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.Lock, contentDescription = null, modifier = Modifier.size(16.dp))
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(viewModel.strings.get("lockControls", "Lock Controls"))
                        }
                    },
                    onClick = {
                        viewModel.lockControls()
                        showMoreMenu = false
                    }
                )

                // Technical Info
                DropdownMenuItem(
                    text = {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(Icons.Default.Info, contentDescription = null, modifier = Modifier.size(16.dp))
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(viewModel.strings.get("technicalInfo", "Technical Info"))
                        }
                    },
                    onClick = {
                        viewModel.showTechnicalInfo = !viewModel.showTechnicalInfo
                        showMoreMenu = false
                    }
                )
            }
        }
    }
}

fun formatTimeSec(seconds: Double): String {
    val total = seconds.toInt().coerceAtLeast(0)
    val hrs = total / 3600
    val mins = (total % 3600) / 60
    val secs = total % 60
    return if (hrs > 0) {
        String.format("%d:%02d:%02d", hrs, mins, secs)
    } else {
        String.format("%d:%02d", mins, secs)
    }
}
